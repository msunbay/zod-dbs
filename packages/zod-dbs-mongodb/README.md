# zod-dbs-mongodb

MongoDB provider for zod-dbs.

## Notes

- Requires `mongodb` as a peer dependency.
- Enum types are not yet supported, but can be added to the generated schemas using hooks.

## Installation

```bash
npm install zod-dbs-mongodb
```

> Note: This package has a peer dependency on `mongodb`. If you don't already have it, you might need to install it (depending on your environment, package manager, etc):

```bash
npm install mongodb
```

## CLI Usage

```bash
npx zod-dbs --provider mongodb \
  --host <host> \
  --user <user> --password <password> \
  --database <db>
```

### Provider Specific CLI Options

| Option                       | Description                                                          | Required |
| ---------------------------- | -------------------------------------------------------------------- | -------- |
| `--connection-string <conn>` | Full database connection string (overrides other connection options) |          |
| `--host <host>`              | MongoDB host (default: `localhost`)                                  |          |
| `--port <port>`              | MongoDB port (default: `27017`)                                      |          |
| `--database <db>`            | MongoDB database name                                                |          |
| `--sample-size <num>`        | Number of documents to sample per collection (default: 50)           |          |
| `--direct-connection`        | Use a direct connection (if needed)                                  |          |
| `--replica-set <name>`       | Name of the replica set (if any)                                     |          |

### Configuration File Example

```ts
import { ZodDbsCliConfig } from 'zod-dbs-cli';

// Import needed to load the provider specific configuration types.
import 'zod-dbs-mongodb';

const config: ZodDbsCliConfig = {
  provider: 'mongodb',
  replicaSet: '<replicaSet>', // optional
  directConnection: true, // optional
  sampleSize: 100, // optional default is 50
};

export default config;
```

## Programmatic Usage

```ts
import { generateZodSchemas } from 'zod-dbs';
import { createProvider } from 'zod-dbs-mongodb';

await generateZodSchemas({
  provider: createProvider(),
  config: {
    host: '<host>',
    user: '<user>',
    password: '<password>',
    token: '<token>',
    database: '<db>',
    replicaSet: '<replicaSet>', // optional
    directConnection: true, // optional,
    sampleSize: 50, // optional
  },
});
```
