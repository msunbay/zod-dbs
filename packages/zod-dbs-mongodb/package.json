{
  "name": "zod-dbs-mongodb",
  "version": "1.0.0",
  "description": "MongoDB provider for zod-dbs",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test:types": "tsc --noEmit"
  },
  "author": "msunbay",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msunbay/zod-dbs.git",
    "directory": "packages/zod-dbs-mongodb"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "zod-dbs-core": "workspace:*"
  },
  "peerDependencies": {
    "mongodb": ">=5"
  },
  "devDependencies": {
    "mongodb": "^6.8.0"
  }
}
