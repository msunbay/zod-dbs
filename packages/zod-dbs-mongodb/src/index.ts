export { MongoDbProvider } from './MongoDbProvider.js';
export { createProvider } from './provider.js';
export { createClient } from './client.js';
export type * from './types.js';
