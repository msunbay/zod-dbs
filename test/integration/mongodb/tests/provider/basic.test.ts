import { MongoDbProvider } from 'zod-dbs-mongodb';

import { getProviderConfig } from '../../../utils/context.js';

it('returns raw schema column information', async () => {
  const provider = new MongoDbProvider();
  const info = await provider.fetchSchemaInfo(getProviderConfig());

  expect(info).toBeDefined();
  expect(info.length).toBeGreaterThan(0);

  // Remove undefined keys for stable snapshots
  info.forEach((column) => {
    (Object.keys(column) as (keyof typeof column)[]).forEach((key) => {
      if (column[key] === undefined) delete column[key];
    });
  });

  expect(info).toMatchSnapshot('rawColumns');
});

it('returns schema models', async () => {
  const provider = new MongoDbProvider();
  const info = await provider.getSchemaInformation(getProviderConfig());

  expect(info).toBeDefined();
  expect(info.tables).toBeDefined();
  expect(info.tables.length).toBeGreaterThan(0);

  const users = info.tables.find((t) => t.name === 'users');
  expect(users).toBeDefined();
  expect(users!.columns.length).toBeGreaterThan(0);
  expect(users).toMatchSnapshot('usersTable');
});
