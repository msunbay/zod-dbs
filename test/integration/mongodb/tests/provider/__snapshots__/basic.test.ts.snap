// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns raw schema column information > rawColumns 1`] = `
[
  {
    "dataType": "objectid",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "_id",
    "schemaName": "testdb",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "string",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "email",
    "schemaName": "testdb",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "date",
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "name": "createdAt",
    "schemaName": "testdb",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "object",
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "name": "profile",
    "objectDefinition": {
      "columns": [
        {
          "dataType": "string",
          "defaultValue": undefined,
          "description": undefined,
          "enumValues": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": false,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "minLen": undefined,
          "name": "displayName",
          "tableName": "users",
          "tableType": "table",
          "zodType": "string",
        },
        {
          "dataType": "int",
          "defaultValue": undefined,
          "description": undefined,
          "enumValues": undefined,
          "isArray": false,
          "isEnum": false,
          "isNullable": false,
          "isReadOptional": true,
          "isSerial": false,
          "isWritable": true,
          "isWriteOptional": true,
          "maxLen": undefined,
          "minLen": undefined,
          "name": "age",
          "tableName": "users",
          "tableType": "table",
          "zodType": "int",
        },
      ],
      "name": "users.profile",
      "type": "object",
    },
    "schemaName": "testdb",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "array",
    "isArray": true,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "name": "tags",
    "schemaName": "testdb",
    "tableName": "users",
    "tableType": "table",
  },
]
`;

exports[`returns schema models > usersTable 1`] = `
{
  "columns": [
    {
      "dataType": "objectid",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "_id",
      "objectDefinition": undefined,
      "schemaName": "testdb",
      "tableName": "users",
      "tableType": "table",
      "zodType": "string",
    },
    {
      "dataType": "string",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "email",
      "objectDefinition": undefined,
      "schemaName": "testdb",
      "tableName": "users",
      "tableType": "table",
      "zodType": "string",
    },
    {
      "dataType": "date",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "createdAt",
      "objectDefinition": undefined,
      "schemaName": "testdb",
      "tableName": "users",
      "tableType": "table",
      "zodType": "date",
    },
    {
      "dataType": "object",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "profile",
      "objectDefinition": {
        "columns": [
          {
            "dataType": "string",
            "defaultValue": undefined,
            "description": undefined,
            "enumValues": undefined,
            "isArray": false,
            "isEnum": false,
            "isNullable": false,
            "isReadOptional": true,
            "isSerial": false,
            "isWritable": true,
            "isWriteOptional": true,
            "maxLen": undefined,
            "minLen": undefined,
            "name": "displayName",
            "tableName": "users",
            "tableType": "table",
            "zodType": "string",
          },
          {
            "dataType": "int",
            "defaultValue": undefined,
            "description": undefined,
            "enumValues": undefined,
            "isArray": false,
            "isEnum": false,
            "isNullable": false,
            "isReadOptional": true,
            "isSerial": false,
            "isWritable": true,
            "isWriteOptional": true,
            "maxLen": undefined,
            "minLen": undefined,
            "name": "age",
            "tableName": "users",
            "tableType": "table",
            "zodType": "int",
          },
        ],
        "name": "users.profile",
        "type": "object",
      },
      "schemaName": "testdb",
      "tableName": "users",
      "tableType": "table",
      "zodType": "object",
    },
    {
      "dataType": "array",
      "defaultValue": undefined,
      "description": undefined,
      "enumValues": undefined,
      "isArray": true,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "minLen": undefined,
      "name": "tags",
      "objectDefinition": undefined,
      "schemaName": "testdb",
      "tableName": "users",
      "tableType": "table",
      "zodType": "any",
    },
  ],
  "name": "users",
  "schemaName": "testdb",
  "type": "table",
}
`;
